hexkit/protocols/dao.py

```diff
@@ -0,0 +1,415 @@
+        """Create a new resource.
+
+        Args:
+            dto:
+                Resource content as a pydantic-based data transfer object including the
+                resource ID.
+
+        Raises:
+            ResourceAlreadyExistsError:
+                when a resource with the id specified in the dto does already exist.
+        """
+        ...
+
+    def upsert(self, dto: Dto) -> None:
+        """Update the provided resource if it already exists, create it otherwise.
+
+        Args:
+            dto:
+                Resource content as a pydantic-based data transfer object including the
+                resource ID.
+        """
+        ...
+
+
+class DaoFactoryProtcol(ABC):
+    """A Protocol describing a factory to produce Data Access Objects (DAO) objects when
+    providing a Data Transfer Objetct (DTO) class.
+    """
+
+    class IdFieldNotFoundError(ValueError):
+        """Raised when the dto_model did not contain the expected id_field."""
+
+    class CreationModelInvalidError(ValueError):
+        """Raised when the DtoCreationModel was invalid in relation to the main
+        DTO model."""
+
+    class IndexFieldsInvalidError(ValueError):
+        """Raised when providing an invalid list of fields to index."""
+
+    @classmethod
+    def _validate_dto_model_id(cls, *, dto_model: type[Dto], id_field: str) -> None:
+        """Checks whether the dto_model contains the expected id_field.
+        Raises otherwises."""
+
+        if id_field not in dto_model.schema()["properties"].keys():
+            raise cls.IdFieldNotFoundError()
+
+    @classmethod
+    def _validate_dto_creation_model(
+        cls,
+        *,
+        dto_model: type[Dto],
+        dto_creation_model: Optional[type[DtoCreation]],
+        id_field: str,
+    ) -> None:
+        """Checks that the dto_creation_model has identical fields than the dto_model
+        except missing the ID. Raises otherwises."""
+
+        if dto_creation_model is None:
+            return
+
+        expected_properties = deepcopy(dto_model.schema()["properties"])
+        # (the schema method returns an attribute of the class, making a copy to not
+        # alter the class)
+        del expected_properties[id_field]
+        observed_properties = dto_creation_model.schema()["properties"]
+
+        if observed_properties != expected_properties:
+            raise cls.CreationModelInvalidError()
+
+    @classmethod
+    def _validate_fields_to_index(
+        cls,
+        *,
+        dto_model: type[Dto],
+        fields_to_index: Optional[Sequence[str]],
+    ) -> None:
+        """Checks that all index fields are present in the dto_model. Raises otherwises."""
+
+        if fields_to_index is None:
+            return
+
+        existing_fields = dto_model.schema()["properties"].keys()
+
+        for field in fields_to_index:
+            if field not in existing_fields:
+                raise cls.IndexFieldsInvalidError(f"Field '{field}' not in DTO model.")
+
```

Comment:
Alternatively:

```py
        if not all(field in existing_fields for field in fields_to_index):
```

or:

```py
        if not set(fields_to_index).issubset(existing_fields):
```

or:

```py
        if not set(fields_to_index) <= set(existing_fields):
```

Reply:
nice, thank you, all of them are much better. I am going for the 2nd one.

Reply:
✅